src/storage/predicate.rs

```diff
@@ -16,104 +15,34 @@ use fuel_types::{Address, AssetId, Bytes32, ContractId, Salt};
 #[derive(Debug, Default, Clone, Copy)]
 pub struct PredicateStorage;
 
-impl Storage<ContractId, Contract> for PredicateStorage {
+impl<Type: Mappable> StorageInspect<Type> for PredicateStorage {
```

Comment:
The storage predicate is a special case of a storage that shouldn't provide a functionality. Since we are refactoring the storage, do you think we can come with a more robust solution to that problem? Its ofc not a blocker for this PR, but its a tech-debt that will have to be addressed either now or later

Reply:
But I guess the bytecode of the predicate may contain the opcodes with access to the storage? (If the user compiled it without using or modified sway?) And in these cases, we need to return an error. So the current implementation is correct.

Reply:
Yes, the proposed change extends the previous implementation so we end up with the same behavior (e.g. returning an error if an access to the storage is attempted).

However, that implementation was a quick&dirty way to have a storage that will also "support predicates" because originally the storage was a requirement for the interpreter. But then we have a mode of the interpreter that shouldn't implement anything related to the storage (i.e. predicate mode) so it will basically just error anything that comes.

If a predicate ends up calling a storage function, then that is a bug in the VM. We should halt any instance of a contract opcode instead of attempting to call the storage (and failing). In other words; yes, it is correct to say that we should error for any predicate verification, but it was implemented like that just to avoid the increased work of making the storage optional for the interpreter.

But as mentioned, this is not really a blocker for this PR, it was just mentioned as window of opportunity to improve things further.

Reply:
Do you have a separate check in the `fuel-vm` that the byte code of the predicate doesn't contain any storage opcode?